core/src/bls_sigverify/bls_sigverifier.rs

```diff
@@ -0,0 +1,1565 @@
+        let mut consensus_metrics_to_send = Vec::new();
+        let mut last_voted_slots: HashMap<Pubkey, Slot> = HashMap::new();
+
+        let root_bank = self.sharable_banks.root();
+        if self.last_checked_root_slot < root_bank.slot() {
+            self.last_checked_root_slot = root_bank.slot();
+            self.verified_certs
+                .write()
+                .unwrap()
+                .retain(|cert| cert.slot() > root_bank.slot());
+            self.vote_payload_cache
+                .write()
+                .unwrap()
+                .retain(|vote, _| vote.slot() > root_bank.slot());
+        }
+
+        for mut packet in batches.iter_mut().flatten() {
+            self.stats.received.fetch_add(1, Ordering::Relaxed);
+            if packet.meta().discard() {
+                self.stats
+                    .received_discarded
+                    .fetch_add(1, Ordering::Relaxed);
+                continue;
+            }
+
+            let message: ConsensusMessage = match packet.deserialize_slice(..) {
+                Ok(msg) => msg,
+                Err(_) => {
+                    self.stats
+                        .received_malformed
+                        .fetch_add(1, Ordering::Relaxed);
+                    packet.meta_mut().set_discard(true);
+                    continue;
+                }
+            };
+
+            match message {
+                ConsensusMessage::Vote(vote_message) => {
+                    // Missing epoch states
+                    let Some(key_to_rank_map) =
+                        get_key_to_rank_map(&root_bank, vote_message.vote.slot())
+                    else {
+                        self.stats
+                            .received_no_epoch_stakes
+                            .fetch_add(1, Ordering::Relaxed);
+                        packet.meta_mut().set_discard(true);
+                        continue;
+                    };
+
+                    // Invalid rank
+                    let Some((solana_pubkey, bls_pubkey)) =
+                        key_to_rank_map.get_pubkey(vote_message.rank.into())
+                    else {
+                        self.stats.received_bad_rank.fetch_add(1, Ordering::Relaxed);
+                        packet.meta_mut().set_discard(true);
+                        continue;
+                    };
+
+                    // Capture votes received metrics before old messages are potentially discarded below.
+                    let slot = vote_message.vote.slot();
+                    if vote_message.vote.is_notarization_or_finalization() {
+                        let existing_slot = last_voted_slots.entry(*solana_pubkey).or_insert(slot);
+                        *existing_slot = (*existing_slot).max(slot);
+                    }
+                    consensus_metrics_to_send.push(ConsensusMetricsEvent::Vote {
+                        id: *solana_pubkey,
+                        vote: vote_message.vote,
```

Comment:
I think we only actually need `VoteType` here. Could save some bits

Reply:
This does involve changing consensus_metrics interface, so maybe we can do that in a follow up PR? @akhi3030 what do you think?